src/vm/core.zig

```diff
@@ -383,19 +389,110 @@ pub const CairoVM = struct {
+/// The operation is allowed between:
+/// 1. A felt and another felt.
+/// 2. A felt and a relocatable.
+/// Adding two relocatables is forbidden.
+/// # Arguments
+/// - `op_0`: The operand 0.
+/// - `op_1`: The operand 1.
+/// # Returns
+/// - `MaybeRelocatable`: The result of the operation or an error.
+pub fn addOperands(
+    op_0: MaybeRelocatable,
+    op_1: MaybeRelocatable,
+) CairoVMError!MaybeRelocatable {
+    // Both operands are relocatables, operation forbidden
+    if (op_0.isRelocatable() and op_1.isRelocatable()) {
+        return error.AddRelocToRelocForbidden;
+    }
+
+    // One of the operands is relocatable, the other is felt
+    if (op_0.isRelocatable() or op_1.isRelocatable()) {
+        // Determine which operand is relocatable and which one is felt
+        const reloc_op = if (op_0.isRelocatable()) op_0 else op_1;
+        const felt_op = if (op_0.isRelocatable()) op_1 else op_0;
+
+        var reloc = try reloc_op.tryIntoRelocatable();
+        var felt = try felt_op.tryIntoFelt();
+
+        // Add the felt to the relocatable's offset
+        try reloc.addFeltInPlace(felt);
+
+        return relocatable.newFromRelocatable(reloc);
+    }
+
+    // Both operands are felts
+    const op_0_felt = try op_0.tryIntoFelt();
+    const op_1_felt = try op_1.tryIntoFelt();
+
+    // Add the felts and return as a new felt wrapped in a relocatable
+    return relocatable.fromFelt(op_0_felt.add(op_1_felt));
+}
+
+/// Compute the product of two operands op 0 and op 1.
+/// # Arguments
+/// - `op_0`: The operand 0.
+/// - `op_1`: The operand 1.
+/// # Returns
+/// - `MaybeRelocatable`: The result of the operation or an error.
+pub fn mulOperands(
+    op_0: MaybeRelocatable,
+    op_1: MaybeRelocatable,
+) CairoVMError!MaybeRelocatable {
+    // At least one of the operands is relocatable
+    if (op_0.isRelocatable() or op_1.isRelocatable()) {
+        return CairoVMError.MulRelocForbidden;
+    }
+
+    const op_0_felt = try op_0.tryIntoFelt();
+    const op_1_felt = try op_1.tryIntoFelt();
+
+    // Multiply the felts and return as a new felt wrapped in a relocatable
+    return relocatable.fromFelt(op_0_felt.mul(op_1_felt));
+}
+
 // *****************************************************************************
 // *                       CUSTOM TYPES                                        *
 // *****************************************************************************
 
 /// Represents the operands for an instruction.
 const OperandsResult = struct {
-    dst: relocatable.MaybeRelocatable,
-    res: ?relocatable.MaybeRelocatable,
-    op_0: relocatable.MaybeRelocatable,
-    op_1: relocatable.MaybeRelocatable,
-    dst_addr: relocatable.MaybeRelocatable,
-    op_0_addr: relocatable.MaybeRelocatable,
-    op_1_addr: relocatable.MaybeRelocatable,
+    dst: MaybeRelocatable,
+    res: ?MaybeRelocatable,
+    op_0: MaybeRelocatable,
+    op_1: MaybeRelocatable,
+    dst_addr: relocatable.Relocatable,
```

Comment:
Here as you import a `MaybeRelocatable` constant type at the beginning of the file, I think we can also import `Relocatable` as `relocatable.Relocatable`

Reply:
done